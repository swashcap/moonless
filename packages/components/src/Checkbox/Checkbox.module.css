/** Checkbox: Label */

.CheckboxLabel {
  display: flex;
}

/** Checkbox: Indicator */

.CheckboxIndicator {
  align-items: center;
  background: var(--theme-color-secondary-background);
  border-color: var(--theme-color-secondary-border);
  border-radius: var(--theme-size-medium-border-radius);
  border-style: solid;
  border-width: var(--theme-size-medium-border-width);
  color: var(--theme-color-secondary-text);
  cursor: pointer;
  display: flex;
  height: 1.5rem;
  justify-content: center;
  margin-right: var(--space-100);
  width: 1.5rem;
}

.CheckboxIndicator:hover,
:global(.storybook-hover) .CheckboxIndicator {
  background: var(--theme-color-secondary-hover-background);
  border-color: var(--theme-color-secondary-hover-border);
  color: var(--theme-color-secondary-hover-text);
}

.CheckboxIndicator:focus,
:global(.storybook-focus) .CheckboxIndicator,
.CheckboxIndicator:active,
:global(.storybook-active) .CheckboxIndicator {
  background: var(--theme-color-secondary-active-background);
  border-color: var(--theme-color-secondary-active-border);
  color: var(--theme-color-secondary-active-text);
}

.CheckboxIndicatorIcon {
  opacity: 0;
}

/** Checkbox: Indicator: Checked, Indeterminate */

/* stylelint-disable no-descending-specificity */
input:checked ~ .CheckboxIndicator,
input:indeterminate ~ .CheckboxIndicator {
  background: var(--theme-color-primary-background);
  border-color: var(--theme-color-primary-border);
  color: var(--theme-color-primary-text);
}

input:checked ~ .CheckboxIndicator:hover,
:global(.storybook-hover) input:checked ~ .CheckboxIndicator {
  background: var(--theme-color-primary-hover-background);
  border-color: var(--theme-color-primary-hover-border);
  color: var(--theme-color-primary-hover-text);
}

input:checked ~ .CheckboxIndicator:focus,
:global(.storybook-focus) input:checked ~ .CheckboxIndicator,
input:checked ~ .CheckboxIndicator:active,
:global(.storybook-active) input:checked ~ .CheckboxIndicator {
  background: var(--theme-color-primary-active-background);
  border-color: var(--theme-color-primary-active-border);
  color: var(--theme-color-primary-active-text);
}

input:indeterminate ~ .CheckboxIndicator .CheckboxIndicatorIcon,
input:checked ~ .CheckboxIndicator .CheckboxIndicatorIcon {
  opacity: 1;
}

/** Checkbox: Indicator: Disabled */

input[disabled] ~ .CheckboxIndicator {
  background: var(--theme-color-secondary-disabled-background);
  border-color: var(--theme-color-secondary-disabled-border);
  color: var(--theme-color-secondary-disabled-text);
  cursor: default;
}

input[disabled]:checked ~ .CheckboxIndicator {
  background: var(--theme-color-primary-disabled-background);
  border-color: var(--theme-color-primary-disabled-border);
  color: var(--theme-color-primary-disabled-text);
}
/* stylelint-enable no-descending-specificity */

/** Checkbox: Error */

.CheckboxError {
  margin-top: var(--space-100);
}
