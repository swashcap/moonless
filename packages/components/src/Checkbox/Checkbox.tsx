import { forwardRef, useEffect, useRef } from 'react';
import { useId } from 'moonless-utils';

import styles from './Checkbox.module.css';
import visuallyHiddenStyles from '../VisuallyHidden/VisuallyHidden.module.css';
import { CheckIcon, MinusIcon } from '../Icon/Icon';
import { ControlError } from '../private/ControlError/ControlError';

export interface CheckboxProps
  extends React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLDivElement>,
    HTMLDivElement
  > {
  checked?: boolean;
  disabled?: boolean;
  error?: React.ReactNode;
  id?: string;
  indeterminate?: boolean;
  label?: React.ReactNode;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

export const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(
  (
    {
      checked,
      disabled,
      error,
      id: idProp,
      indeterminate,
      label,
      onChange,
      ...rest
    },
    ref
  ) => {
    const fallbackId = useId('checkbox-');
    const errorId = useId('checkbox-error-');
    const localRef = useRef<HTMLInputElement | null>();

    useEffect(() => {
      if (
        localRef.current &&
        localRef.current.indeterminate !== indeterminate
      ) {
        localRef.current.indeterminate = !!indeterminate;
      }
    }, [indeterminate, localRef]);

    const id = idProp ?? fallbackId;

    return (
      <div {...rest}>
        <label className={styles.CheckboxLabel} htmlFor={id}>
          <input
            aria-describedby={error ? errorId : undefined}
            aria-invalid={error ? 'true' : undefined}
            checked={checked}
            className={visuallyHiddenStyles.VisuallyHidden}
            disabled={disabled}
            id={id}
            onChange={onChange}
            ref={(element) => {
              if (typeof ref === 'function') {
                ref(element);
              } else if (typeof ref === 'object' && ref) {
                ref.current = element;
              }

              localRef.current = element;
            }}
            type="checkbox"
          />
          <span aria-hidden="true" className={styles.CheckboxIndicator}>
            {indeterminate ? (
              <MinusIcon className={styles.CheckboxIndicatorIcon} />
            ) : (
              <CheckIcon className={styles.CheckboxIndicatorIcon} />
            )}
          </span>
          {label}
        </label>
        {!!error && (
          <ControlError className={styles.CheckboxError} id={errorId}>
            {error}
          </ControlError>
        )}
      </div>
    );
  }
);
