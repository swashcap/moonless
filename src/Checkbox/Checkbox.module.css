/** Checkbox: Label */

.CheckboxLabel {
  display: flex;
}

/** Checkbox: Indicator */

.CheckboxIndicator {
  align-items: center;
  background: var(--color-white);
  border-radius: 4px;
  border: 1px solid var(--color-black);
  cursor: pointer;
  display: flex;
  height: 1.5rem;
  justify-content: center;
  margin-right: var(--space-100);
  width: 1.5rem;
}

/** Checkbox: Indicator: Checked */

input:checked ~ .CheckboxIndicator {
  background: var(--color-blue);
  border-color: var(--color-blue);
  color: var(--color-white);
}

/** Checkbox: Indicator: Disabled */

input[disabled] ~ .CheckboxIndicator {
  border-color: var(--color-gray);
  color: var(--color-gray);
}

input[disabled]:checked ~ .CheckboxIndicator {
  background: var(--color-gray);
  color: var(--color-white);
}

/** Checkbox: Error */

.CheckboxError {
  color: var(--color-red);
  font-size: var(--typography-font-size-100);
  line-height: var(--typography-line-height-100);
  margin-top: var(--space-100);
  width: 100%;
}
