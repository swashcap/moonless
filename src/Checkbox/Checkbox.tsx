import React from 'react';

import styles from './Checkbox.module.css';
import { CheckIcon } from '../Icon/Icon';
import { VisuallyHidden } from '../VisuallyHidden/VisuallyHidden.module.css';
import { useId } from '../utils/useId';

export interface CheckboxProps
  extends React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLDivElement>,
    HTMLDivElement
  > {
  checked?: boolean;
  disabled?: boolean;
  error?: React.ReactNode;
  id?: string;
  label?: React.ReactNode;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => any;
}

export const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>(
  ({ checked, disabled, error, id: idProp, label, onChange, ...rest }, ref) => {
    const fallbackId = useId('checkbox-');
    const errorId = useId('checkbox-error-');

    const id = idProp ?? fallbackId;

    return (
      <div {...rest}>
        <label className={styles.CheckboxLabel} htmlFor={id}>
          <input
            aria-describedby={error ? errorId : undefined}
            aria-invalid={error ? 'true' : undefined}
            checked={checked}
            className={VisuallyHidden}
            disabled={disabled}
            id={id}
            onChange={onChange}
            ref={ref}
            type="checkbox"
          />
          <span className={styles.CheckboxIndicator}>
            <CheckIcon aria-hidden="true" />
          </span>
          {label}
        </label>
        {!!error && (
          <span className={styles.CheckboxError} id={errorId}>
            {error}
          </span>
        )}
      </div>
    );
  }
);
