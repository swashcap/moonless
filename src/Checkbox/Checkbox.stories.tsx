import React from 'react';

import { Checkbox } from './Checkbox';

export default {
  component: Checkbox,
  title: 'Checkbox',
};

export const Default = () => <Checkbox />;

export const Checked = () => {
  const [checked, setChecked] = React.useState(true);

  return <Checkbox checked={checked} onChange={() => setChecked(!checked)} />;
};

export const Disabled = () => (
  <>
    <Checkbox disabled />
    <Checkbox className="storybook-focus" disabled />
    <Checkbox className="storybook-hover" disabled />
    <Checkbox className="storybook-active" disabled />
  </>
);

export const WithError = () => <Checkbox error="Error" />;

export const WithLabel = () => <Checkbox label="Label" />;
