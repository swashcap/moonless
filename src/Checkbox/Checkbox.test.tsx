import React from 'react';
import { render } from '@testing-library/react';

import { Checkbox } from './Checkbox';

describe('Checkbox', () => {
  test('It should render.', () => {
    const { container } = render(<Checkbox />);

    expect(container).toMatchSnapshot();
  });
});
