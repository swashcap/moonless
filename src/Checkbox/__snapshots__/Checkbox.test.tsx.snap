// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox It should render. 1`] = `
<div>
  <div
    class="Checkbox"
  >
    <label
      class="CheckboxLabel"
      for="checkbox-1"
    >
      <input
        class="VisuallyHidden"
        id="checkbox-1"
        type="checkbox"
      />
      <span
        class="CheckboxIndicator"
      >
        <svg
          aria-hidden="true"
          class="Icon"
          viewBox="0 0 12 12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2.5 6.5l2.5 3 4.5-6"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </span>
    </label>
  </div>
</div>
`;
